src/devices/Hcsr04/Hcsr04.cs

```diff
@@ -29,14 +29,14 @@ public class Hcsr04 : IDisposable
         /// <summary>
         /// Creates a new instance of the HC-SCR04 sonar.
         /// </summary>
+        /// <param name="gpioController">GPIO controller related with the pins</param>
         /// <param name="triggerPin">Trigger pulse input.</param>
         /// <param name="echoPin">Trigger pulse output.</param>
-        /// <param name="pinNumberingScheme">Pin Numbering Scheme</param>
-        public Hcsr04(int triggerPin, int echoPin, PinNumberingScheme pinNumberingScheme = PinNumberingScheme.Logical)
+        public Hcsr04(GpioController gpioController, int triggerPin, int echoPin)
```

Comment:
Now you need to keep track whether we created the controller ourselves or not, because like this, `Dispose()` will clean it up in either case.

Reply:
@pgrawehr nah, per our guidelines ownership of GPIO controller belongs to this binding:
https://github.com/dotnet/iot/blob/master/Documentation/Devices-conventions.md#lifetime-management-or-what-should-be-disposed-by-what

if we want to keep track of this we should add extra argument `shouldDispose`

Reply:
I know, but then we need this parameter, because otherwise the usage of this is quite problematic still. If we (say) provide a Mcp23xxx as GpioController, we probably don't want it to be disposed when disposing the Hcsr (as there might be other devices still attached to the controller we want to still use or shutdown gracefuly as well using their bindings).

Reply:
@pgrawehr there should be one instance of GpioController per device, it should not be shared.

Reply:
That only works as long as several GpioController instances can be created for the same physical device and Disposing one of them does not affect the others. This may work for the default Drivers (Raspi, SysFs), but will fail for things like a GpioController created from an MCP23017 (because we generally cannot create two instances of the same I2C device - the bindings would interfere with each other). Note also that with "One GpioController per device" approach, there's no way of blocking the user from doing stupid things like assigning the same pin to several devices at once.

Of course, there could be other ways of working around this issue, like providing Clone() operations that create "Slave" instances which internally reference the master component and do reference counting, but that seems a lot more complex, both in implementation and in usage. 

The `Board.CreateGpioController()`-Approach could be an option as well (then the concrete implementation on an MCP would probably just do nothing on Dispose as that is handled by the master board component). This would probably introduce a breaking change on MCP usage, though (or at least make the current usage variant deprecated).

Reply:
@pgrawehr I'll leave this is as is for now and we should start some conversation in separate issue about this. If this is an issue for all devices we should fix it globally and adjust guidelines or fix MCP23xxx to implement GpioDriver correctly.

Do you mind creating an issue about this and creating any sample code which shows the specific issue?

My current understanding is that there should be a way of creating GpioController multiple times but if there is some architectural limitation (i.e. expanders using I2C or whatever) then we should fix the problem not the symptoms

Reply:
Ok. We clearly have a conceptual problem there. As stated above, there might be different solutions to it.
I'll collect information on the problem, write it up and then try to actually reproduce it. Might take some time tough, since I'm having quite a bunch of open ends right now (aka too many open branches). Also I'm missing a Socket to get my MCP properly attached to the breadboard and ordering will take some time.